### PR TITLE
Eng 9360 - conflict detection and resolution

### PR DESCRIPTION
**Pull request by luyangco**
Passed manual test.

### CODE REVIEW

File: src/ee/execution/JNITopend.cpp [hunk 1]
```diff
@@ -476,52 +476,84 @@ int64_t JNITopend::pushDRBuffer(int32_t partitionId, StreamBlock *block) {
...
+        existingArrayForDelete = boost::shared_ptr<char>(new char[serializeSize]);
+        ReferenceSerializeOutput conflictSerializeOutput(existingArrayForDelete.get(), serializeSize);
+        existingTableForDelete->serializeToWithoutTotalSize(conflictSerializeOutput);
+
+        existingRowsBufferForDelete = m_jniEnv->NewDirectByteBuffer(static_cast<void*>(existingArrayForDelete.get()),
+                                                    static_cast<int32_t>(serializeSize));
+        if (existingRowsBufferForDelete == NULL) {
+            m_jniEnv->ExceptionDescribe();
+           throw std::exception();
+        }
```
Comment (nshi) [inline src/ee/execution/JNITopend.cpp]: This block of code used to be wrapped inside a method, any reason it's being duplicated now?

File: src/ee/storage/BinaryLogSink.cpp [hunk 2]
```diff
@@ -459,6 +477,7 @@ void BinaryLogSink::exportDRConflict(Table *exportTable, bool resolved, TempTabl
     if (expectedTableForDelete) {
         TableIterator iterator = expectedTableForDelete->iterator();
         while (iterator.next(tempTuple)) {
+            //TODO: decision column on expected tuple is meanless.
```
Comment (nshi) [inline src/ee/storage/BinaryLogSink.cpp]: Why is this a TODO? It reads just like a normal comment.

File: src/frontend/org/voltdb/PartitionDRGateway.java [hunk 3]
```diff
@@ -45,9 +45,9 @@
         NEW_ROW
     }
 
-    public static enum DRRowDecision {
-        KEEP_ROW,
-        DELETE_ROW;
+    public static enum DRConflictResolutionFlag {
+        ACCEPT_CHANGE,
+        CONVERGENT
```
Comment (nshi) [inline src/frontend/org/voltdb/PartitionDRGateway.java]: It feels weird that these two things are in the same enum. Don't they have overlaps?

File: tests/ee/storage/table_and_indexes_test.cpp [hunk 4]
```diff
@@ -417,7 +417,7 @@ TEST_F(TableAndIndexTest, DrTest) {
     *reinterpret_cast<int32_t*>(&data.get()[startPos]) = htonl(static_cast<int32_t>(sb->offset()));
     drStream.m_enabled = false;
     districtTable->setDR(false);
-    sink.apply(&data[startPos], tables, &pool, mockEngine);
+    sink.apply(&data[startPos], tables, &pool, mockEngine, 0);
```
Comment (nshi) [inline tests/ee/storage/table_and_indexes_test.cpp]: The table_and_indexes_test test suite uses cluster ID 0 as the local cluster ID. The remote cluster ID passed in here should different from that.

File: src/frontend/org/voltdb/PartitionDRGateway.java (hunk 3 above)
Comment (nshi) [inline src/frontend/org/voltdb/PartitionDRGateway.java]: After reading the code that uses this enum, I think I understand it now. These enums are indices, not values.

File: src/ee/storage/BinaryLogSink.cpp (hunk 2 above)
Comment (luyangco) [inline src/ee/storage/BinaryLogSink.cpp]: I'll remove TODO mark, it is just a comment.

File: src/ee/execution/JNITopend.cpp (hunk 1 above)
Comment (luyangco) [inline src/ee/execution/JNITopend.cpp]: There is just my laziness not encapsulate them into a function, I'll wrap them.

**Comment by nshi** [comment]
Squash merged externally

**Pull request closed without merging**